nasl/nasl_packet_forgery.c

```diff
@@ -724,6 +732,182 @@ get_tcp_element (lex_ctxt *lexic)
+  if (options == NULL)
+    return;
+
+  opt_kind = (uint8_t *) options;
+
+  while (*opt_kind != 0)
+    {
+      switch (*opt_kind)
+        {
+        case TCPOPT_MAXSEG:
+          tcp_all_options->mss.kind = *opt_kind;
+          tcp_all_options->mss.len = *(opt_kind + 1);
+          tcp_all_options->mss.mss = *((uint16_t *) (opt_kind + 2));
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_WINDOW:
+          tcp_all_options->wscale.kind = *opt_kind;
+          tcp_all_options->wscale.len = *(opt_kind + 1);
+          tcp_all_options->wscale.wscale = (uint8_t) * (opt_kind + 2);
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_SACK_PERMITTED:
+          tcp_all_options->sack_perm.kind = *opt_kind;
+          tcp_all_options->sack_perm.len = *(opt_kind + 1);
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_TIMESTAMP:
+          tcp_all_options->tstamp.kind = *opt_kind;
+          tcp_all_options->tstamp.len = *(opt_kind + 1);
+          tcp_all_options->tstamp.tstamp = *((uint32_t *) (opt_kind + 2));
+          tcp_all_options->tstamp.e_tstamp = *((uint32_t *) (opt_kind + 6));
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        case TCPOPT_EOL:
+        case TCPOPT_NOP:
+          opt_kind++;
+          break;
+        case TCPOPT_SACK: // Not supported
+          opt_kind = opt_kind + *(opt_kind + 1);
+          break;
+        default:
+          g_debug ("%s: Unsupported %u tcp option. "
+                   "Not all options are returned",
+                   __func__, *opt_kind);
+          *opt_kind = 0;
+          break;
+        }
+    }
+}
+
+/**
+ * @brief Get a TCP option from an IP datagram if present.
+ * Possible options are:
+ *   TCPOPT_MAXSEG (2), values between 536 and 65535
+ *   TCPOPT_WINDOW (3), with values between 0 and 14
+ *   TCPOPT_SACK_PERMITTED (4), no value required.
+ *   TCPOPT_TIMESTAMP (8), 8 bytes value for timestamp
+ *   and echo timestamp, 4 bytes each one.
+ *
+ * @param[in] lexic   Lexical context of NASL interpreter.
+ * @param[in] tcp     The full IP datagram (IP + TCP).
+ * @param[in] option  Option to get.
+ *
+ * @return  Integer or array given the case..
+ */
+tree_cell *
+get_tcp_option (lex_ctxt *lexic)
+{
+  u_char *packet = (u_char *) get_str_var_by_name (lexic, "tcp");
+  struct ip *ip;
+  int ipsz;
+  struct tcphdr *tcp;
+  char *options;
+  int opt;
+  tree_cell *retc;
+  nasl_array *arr;
+  anon_nasl_var v;
+
+  struct tcp_options *tcp_all_options = NULL;
+
+  ipsz = get_var_size_by_name (lexic, "tcp");
+
+  if (packet == NULL)
+    {
+      nasl_perror (lexic, "%s: No valid 'tcp' argument\n", __func__);
+      return NULL;
+    }
+
+  opt = get_int_var_by_name (lexic, "option", -1);
+  if (opt < 0)
+    {
+      nasl_perror (lexic,
+                   "%s: No options\n."
+                   "Usage: get_tcp_option (tcp:tcp, option: TCPOPT_MAXSEG)",
```

Comment:
`get_tcp_option` probably should be `__func__` as well?

Also note the difference between tcp:tcp (no space) vs. option: TCPOPT_MAXSEG (space after the :).